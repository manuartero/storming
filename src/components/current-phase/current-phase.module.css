.currentPhase {
  /* main grid layout */
  grid-area: current-phase;

  display: grid;
  grid-template-rows: 1fr 6fr;
  grid-template-columns: 1fr;
  grid-template-areas:
    "heading"
    "content";

}

.heading {
  grid-area: heading;
  margin-bottom: 1vh;

  display: flex;
  align-items: center;
  justify-content: center;
}

.content {
  grid-area: content;

  display: flex;
  flex-direction: column;
 
}

.planification {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.action {
}
