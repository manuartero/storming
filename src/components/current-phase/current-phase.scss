.current-phase {
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  overflow: hidden;

  &__heading {
    display: flex;
    flex: 1;
    max-height: 10%;
    align-items: center; // center Y

    border-bottom: 1px solid #ccc;
  }

  &__content {
    display: flex;
    flex-direction: column;
    flex: 1;
    max-height: 90%;
    justify-content: center;

    &__planification {}

    &__action {}
  }
}

.card-silhouette {
  margin: 10px; // FIXME
}
