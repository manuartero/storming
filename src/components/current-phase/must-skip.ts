import { getAvailableTilesForActionCard } from "game-logic/available-tiles";

export function mustSkip({
  activeCard,
  board,
}: {
  activeCard: Card | undefined;
  board: Board;
}) {
  if (activeCard?.cardType === "actionCard") {
    return (
      activeCard.action !== "diplo" &&
      getAvailableTilesForActionCard({ board, activeCard }).length === 0
    );
  }
  return false;
}
