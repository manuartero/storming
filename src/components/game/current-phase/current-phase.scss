.current-phase {
  width: 100%;

  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;

  background-color: rgb(243, 220, 220);

  &__heading {
    height: 100px;
  }

  &__planning {
    height: 100%;

    .card-silhouette {
      margin: 10px; // FIXME
    }
  }

  &__action {
    height: 100%;
  }

}
