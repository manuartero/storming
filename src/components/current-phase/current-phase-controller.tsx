import { useGameContext } from "game-context";
import { isActionCard } from "models/new-card";
import { warnInconsistentState } from "utils/console";
import { CurrentPhase } from "./current-phase";
import { mustSkip } from "./must-skip";

export function CurrentPhaseController() {
  const gameContext = useGameContext();

  if (!gameContext.activePlayer) {
    warnInconsistentState("<CurrentPhase />: no active player");
    return <>ERROR</>;
  }

  if (gameContext.phase === "setup") {
    return <div className="current-phase" />;
  }

  return (
    <CurrentPhase
      phase={gameContext.phase}
      activePlayer={gameContext.activePlayer}
      activeCard={gameContext.activeCard}
      nextActionCard={nextNotCommitedActionCard(gameContext.next)}
      futureActionCard={nextNotCommitedActionCard(gameContext.future)}
      mustSkip={mustSkip(gameContext)}
      onSkip={() => {
        gameContext.skip();
      }}
      onSubmitPlan={() => {
        gameContext.submitPlanification();
      }}
      onCleanActionCard={(actions) => {
        gameContext.plan(actions);
      }}
    />
  );
}

function nextNotCommitedActionCard(
  timelineCards: TimelineCard[]
): ActionCard | undefined {
  const card = timelineCards.find(
    (timelineCard) => !timelineCard.commited
  )?.card;
  return card && isActionCard(card) ? card : undefined;
}
