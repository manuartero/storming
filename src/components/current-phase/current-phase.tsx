import c from "classnames";
import { Button } from "elements";
import { Card, CardSilhouette } from "components/cards";

import { logRender } from "utils/console";

import "./current-phase.scss";

type ActionPhaseProps = {
  phase: "action";
  activePlayer: PlayerType;
  activeCard?: Card;
  mustSkip: boolean;
  onSkip: () => void;
};

type PlanningPhaseProps = {
  phase: "planification";
  activePlayer: PlayerType;
  nextAction?: ActionCard;
  futureAction?: ActionCard;
  event?: EventCard; // TODO Event Cards
  onSubmitPlan: () => void;
};

export function CurrentPhase(props: ActionPhaseProps | PlanningPhaseProps) {
  logRender("CurrentPhase");

  return (
    <div className={c("current-phase")}>
      <div className={c("current-phase__heading", props.activePlayer)}>
        <h2>{props.phase}</h2>
      </div>
      <div className="current-phase__content">
        {props.phase === "action" ? (
          <ActionPhase {...props} />
        ) : (
          <PlanningPhase {...props} />
        )}
      </div>
    </div>
  );
}

function ActionPhase({ activeCard, mustSkip, onSkip }: ActionPhaseProps) {
  if (!activeCard) {
    return <>ERROR</>;
  }

  return (
    <div className="current-phase__action">
      <Card card={activeCard} status="active" />
      <Button disabled={!mustSkip} onClick={onSkip}>
        SKIP
      </Button>
    </div>
  );
}

function PlanningPhase({
  activePlayer,
  nextAction,
  futureAction,
  onSubmitPlan,
}: PlanningPhaseProps) {
  const buttonDisabled = !nextAction || !futureAction;
  return (
    <div className="current-phase__planification">
      {nextAction ? <Card card={nextAction} /> : <CardSilhouette card="next" />}
      {futureAction ? (
        <Card card={futureAction} />
      ) : (
        <CardSilhouette card="future" />
      )}
      <Button
        className={c(activePlayer, buttonDisabled && "button--disabled")}
        onClick={onSubmitPlan}
        disabled={buttonDisabled}
      >
        GO
      </Button>
    </div>
  );
}
