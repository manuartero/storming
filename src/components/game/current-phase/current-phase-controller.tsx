import { useGameContext } from "contexts";
import { CurrentPhase } from "./current-phase";
import { mustSkip } from "./must-skip";

/**
 * Renders depends on:
 *  - `useGameContext()`
 *
 * Defines visual current phase from GameContext
 *
 */
export function CurrentPhaseController(): JSX.Element {
  const gameContext = useGameContext();

  return (
    <CurrentPhase
      phase={gameContext.phase}
      mustSkip={mustSkip(gameContext)}
      activeCard={gameContext.activeCard}
      onSkip={() => {
        gameContext.skip();
      }}
    />
  );
}
