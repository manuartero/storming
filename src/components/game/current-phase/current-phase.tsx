import { Button } from "components/common";
import IconCard from "components/game/cards/icon-card"; // TODO: deprecated
import { logRender } from "utils/console";
import { CardSilhouette } from "components/game/cards";

import "./current-phase.scss";

interface Props {
  phase: PhaseType;
  activeCard: Card | undefined;
  mustSkip: boolean;
  onSkip: () => void;
}

export function CurrentPhase({
  phase,
  activeCard,
  mustSkip,
  onSkip,
}: Props): JSX.Element {
  logRender("CurrentPhase");

  return (
    <div className="current-phase">
      <h2 className="current-phase__heading">{phase}</h2>
      {phase === "action" ? (
        <ActionPhase
          activeCard={activeCard}
          mustSkip={mustSkip}
          onSkip={onSkip}
        />
      ) : (
        <PlanningPhase />
      )}
    </div>
  );
}

function ActionPhase({ activeCard, mustSkip, onSkip }: Omit<Props, "phase">) {
  return (
    <div className="current-phase__action">
      <IconCard card={activeCard} />
      <Button disabled={!mustSkip} onClick={onSkip}>
        SKIP
      </Button>
    </div>
  );
}

function PlanningPhase() {
  return (
    <div className="current-phase__planning">
      <CardSilhouette card="next" />
      <CardSilhouette card="future" />
      <Button disabled={true} onClick={() => {}}>
        GO
      </Button>
    </div>
  );
}
