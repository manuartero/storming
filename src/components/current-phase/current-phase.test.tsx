import { render } from "@testing-library/react";
import { CurrentPhase } from "./current-phase";
import { NewCard } from "models/new-card";

describe("<CurrentPhase />", () => {
  test("render: match snapshot - action", () => {
    const { asFragment } = render(
      <CurrentPhase
        phase="action"
        activePlayer="player"
        activeCard={NewCard("move", "player")}
        mustSkip={false}
        onSkip={jest.fn()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("render: match snapshot - planification", () => {
    const { asFragment } = render(
      <CurrentPhase
        phase="planification"
        activePlayer="player"
        onSubmitPlan={jest.fn()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("render: match snapshot - planification with cards", () => {
    const { asFragment } = render(
      <CurrentPhase
        phase="planification"
        activePlayer="player"
        nextAction={NewCard("move", "player")}
        futureAction={NewCard("recruit", "player")}
        onSubmitPlan={jest.fn()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
