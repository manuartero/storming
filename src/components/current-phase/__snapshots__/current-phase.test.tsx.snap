// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<CurrentPhase /> render: section role - action 1`] = `
<section
  aria-label="current phase"
  class="currentPhase"
  role="region"
>
  <div
    class="heading player"
  >
    <h2>
      action
    </h2>
  </div>
  <div
    class="content"
  >
    <article
      aria-disabled="true"
      aria-label="card player_move_1"
      aria-pressed="false"
      class="card active actionCard"
    >
      <div
        class="heading"
      >
        <div
          class="icon"
        />
        <div
          class="title title"
        >
          move
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class="text paragraph"
        >
          <p>
            Move up to two different Units
          </p>
        </div>
        <div
          class="waterMark"
        />
      </div>
    </article>
    <button
      class="button disabled"
    >
      SKIP
    </button>
  </div>
</section>
`;

exports[`<CurrentPhase /> render: section role - planification 1`] = `
<section
  aria-label="current phase"
  class="currentPhase"
  role="region"
>
  <div
    class="heading player"
  >
    <h2>
      planification
    </h2>
  </div>
  <div
    class="content"
  >
    <div
      aria-disabled="true"
      class="card silhouette"
    >
      <div
        class="title title"
      >
        next
      </div>
      <div
        class="text"
      >
        Drag here the card you wish to use next round
      </div>
    </div>
    <div
      aria-disabled="true"
      class="card silhouette"
    >
      <div
        class="title title"
      >
        future
      </div>
      <div
        class="text"
      >
        Drag here the card you wish to use in two rounds
      </div>
    </div>
    <button
      class="button disabled player"
    >
      GO
    </button>
  </div>
</section>
`;

exports[`<CurrentPhase /> render: section role - planification with cards 1`] = `
<section
  aria-label="current phase"
  class="currentPhase"
  role="region"
>
  <div
    class="heading player"
  >
    <h2>
      planification
    </h2>
  </div>
  <div
    class="content"
  >
    <article
      aria-disabled="false"
      aria-label="card player_move_2"
      aria-pressed="false"
      class="card available actionCard"
    >
      <div
        class="heading"
      >
        <div
          class="icon"
        />
        <div
          class="title title"
        >
          move
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class="text paragraph"
        >
          <p>
            Move up to two different Units
          </p>
        </div>
        <div
          class="waterMark"
        />
      </div>
    </article>
    <article
      aria-disabled="false"
      aria-label="card player_recruit_1"
      aria-pressed="false"
      class="card available actionCard"
    >
      <div
        class="heading"
      >
        <div
          class="icon"
        />
        <div
          class="title title"
        >
          recruit
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class="text paragraph"
        >
          <p>
            Recruit units at up to two different Settlements
          </p>
        </div>
        <div
          class="waterMark"
        />
      </div>
    </article>
    <button
      class="button player"
    >
      GO
    </button>
  </div>
</section>
`;
